libsolidity/lsp/LanguageServer.cpp

```diff
@@ -130,34 +130,75 @@ Json::Value semanticTokensLegend()
 	return legend;
 }
 
+class SemanticTokens: public HandlerBase
+{
+public:
+	explicit SemanticTokens(LanguageServer& _languageServer): HandlerBase(_languageServer) {}
+	void initialize(Json::Value const& _clientCapabilities, Json::Value& _replyCapabilities) override;
+	void operator()(MessageID _id, Json::Value const& _args);
+};
+
+void SemanticTokens::initialize(Json::Value const& _clientCapabilities, Json::Value& _replyCapabilities)
+{
+	if (_clientCapabilities["textDocument"]["semanticTokens"])
+	{
+		_replyCapabilities["semanticTokensProvider"]["legend"] = semanticTokensLegend();
+		_replyCapabilities["semanticTokensProvider"]["range"] = false;
+		_replyCapabilities["semanticTokensProvider"]["full"] = true; // XOR requests.full.delta = true
+
+		m_server.registerHandler("textDocument/semanticTokens/full", *this);
+	}
+}
+
+void SemanticTokens::operator()(MessageID _id, Json::Value const& _args)
+{
+	auto const uri = _args["textDocument"]["uri"];
+
+	m_server.compile();
+
+	auto const sourceName = fileRepository().uriToSourceUnitName(uri.as<string>());
+	SourceUnit const& ast = m_server.compilerStack().ast(sourceName);
+	charStreamProvider().charStream(sourceName);
+
+	Json::Value data = SemanticTokensBuilder().build(ast, charStreamProvider().charStream(sourceName));
+
+	Json::Value reply = Json::objectValue;
+	reply["data"] = data;
+
+	client().reply(_id, std::move(reply));
+}
```

Comment:
@cameel Do you want me to do that in exactly this PR? (Because that's a different topic you're addressing here).

Reply:
Not sure I understand. Why is it a different topic? It's this PR that adds the handler so it could just as well add it as a separate file. It's not modifying an existing handler. Unless I'm missing some reason that prevented you from doing this?